hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/RateLimiter.java

```diff
@@ -135,18 +135,32 @@ protected synchronized long getTimeUnitInMillis() {
 
   /**
    * Is there at least one resource available to allow execution?
-   * @return true if there is at least one resource available, otherwise false
+   * @return the waitInterval to backoff, or 0 if execution is allowed
    */
-  public boolean canExecute() {
+  public long canExecute() {
     return canExecute(1);
   }
 
+  /**
+   * Are there enough available resources to allow execution?
+   * @param amount the number of required resources, a non-negative number
+   * @return the waitInterval to backoff, or 0 if execution is allowed
+   */
+  public synchronized long canExecute(final long amount) {
+    assert amount >= 0;
+    long waitInterval = waitInterval(amount);
```

Comment:
It might be good to call isAvailable first, because I think it handles refilling

Reply:
Also since it all happens within a synchronized method not, may not need to check both states below. Basically check is available first, return 0 if true, otherwise return getWaitInterval

Reply:
yeah I agree that that _should_ be okay since it's all synchronized now, though I do kind of like the idea of explicitly not returning a waitInterval of 0 if we know that that's not desired

Reply:
I guess it really doesn't matter though, because the user will interpret 0 the same way regardless of how it's returned

Reply:
are you sure? 😄